src/process.ml

```diff
@@ -30,9 +30,67 @@ let map_result
       | 0 -> Ok (f ())
       | n -> Error n
 
-type std_output_to =
-  | Terminal
-  | File        of Path.t
+module Output = struct
+  type t =
+    { kind     : kind
+    ; fd       : Unix.file_descr Lazy.t
+    ; channel  : out_channel Lazy.t
+    ; mutable status : status
+    }
+
+  and kind =
+    | File of Path.t
+    | Terminal
+
+  and status =
+    | Keep_open
+    | Close_after_exec
+    | Closed
+
+  let terminal oc =
+    let fd = Unix.descr_of_out_channel oc in
+    { kind = Terminal
+    ; fd = lazy fd
+    ; channel = lazy stdout
+    ; status = Keep_open
+    }
+  let stdout = terminal stdout
+  let stderr = terminal stderr
+
+  let file fn =
+    let fd =
+      lazy (Unix.openfile (Path.to_string fn)
```

Comment:
Presumably this has the same issue I was worried about - that by the time this is actually executed we might have changed dir. I wasn't at all convinced it was necessary, we certainly call `Output.fd` before `Scheduler.with_chdir`, but are there any other paths by which cwd might change?

Reply:
Nope, we do all the chdirs at a single place in Dune. I added a comment to make sure we don't move the Output.fd calls inside the `with_chdir`.